### ISSUE TITLE
How can I confirm bounces and clicks?

### ISSUE THREAD

**Issue by erikwestlund** [spatie/mailcoach-support#258]
Hi,

I set everything up with Amazon SES and sent out a few test campaigns.

I used a few different email addresses, including two ones that should bounce.

I'm not seeing any statistics.  I opened them all and get no click statistics. I also do not see any failures with bounces.

When I check the original message I'm not sure I see anything related to tracking either. How can I confirm the beacon image went out?

I also see: " 2 unprocessed webhooks" in debug.

I'm unclear on how to make sure webhooks get processed.  There are two in my webhook_calls table. I don't see which command in the scheduler clears these.

I'm on v3.

**Comment by freekmurze** [comment]
Hi,

Are you using Horizon? And if so, are you sure it is running? Do you see pending jobs there? Could you take a look at the payload of the webhook and see what those webhooks concern?

Did you follow the instructions in our docs on how to setup SES?
https://mailcoach.app/docs/v3/app/mail-configuration/amazon-ses

**Comment by erikwestlund** [comment]
It is possible Horizon wasn't running at the exact moment those webhooks came in.  Would they not be queued then?  

Yes, I followed the SES directions closely and double-checked.  

Here are the payloads:

```
{
   "Type":"SubscriptionConfirmation",
   "Token":...,
   "Message":"You have chosen to subscribe to the topic ....:mailcoach.\nTo confirm the subscription, visit the SubscribeURL included in this message.",
   "TopicArn":".....:mailcoach",
   "MessageId":"db95328c-ba5a-4216-be0f-ef700ca96596",
   "Signature":"....",
   "Timestamp":"2020-10-09T08:45:49.712Z",
   "SubscribeURL":"https://sns.us-east-1.amazonaws.com/?Action=ConfirmSubscription&TopicArn=....",
   "SigningCertURL":"....",
   "SignatureVersion":"1"
}
```

and 


```
{
   "Type":"Notification",
   "Message":"Successfully validated SNS topic for Amazon SES event publishing.",
   "TopicArn":"....:mailcoach",
   "MessageId":"dbd01550-c905-5b8b-8ec1-42aa65766413",
   "Signature":"...",
   "Timestamp":"2020-10-09T18:21:39.864Z",
   "SigningCertURL":"....",
   "UnsubscribeURL":"https://sns.us-east-1.amazonaws.com/?Action=Unsubscribe&SubscriptionArn=....",
   "SignatureVersion":"1"
}
```

**Comment by erikwestlund** [comment]
I think I figured this out:

If you go into the feedback package from Spatie, and AddConfigurationSetHeader file, it has:

```
        if (! config('mail.driver') === 'ses') {
            return;
        }
```

But check out the default config files on Laravel:

https://github.com/laravel/laravel/blob/master/config/mail.php

There is no "driver" field, and there hasn't been since 6.0. See:

https://github.com/laravel/laravel/blob/6.x/config/mail.php (6.0)

vs.

https://github.com/laravel/laravel/blob/7.x/config/mail.php

vs. 

https://github.com/laravel/laravel/blob/master/config/mail.php

This feature for SES will never work without using non-standard config -- which is a workaround for the moment.

**Comment by erikwestlund** [comment]
Pull request submitted.

One other suggestion: Your docs do not show "open" and "clicked" set in the SES set up. I recommend you make a reference to that!

**Comment by freekmurze** [comment]
> One other suggestion: Your docs do not show "open" and "clicked" set in the SES set up. I recommend you make a reference to that!

@riasvdv can you take a look a the docs? I'll take a look at the PR. Thanks @erikwestlund

**Comment by erikwestlund** [comment]
It appears to me that I was actually originally following the directions correctly, because I bet you guys are programmatically turning on click and open checking, and so you leave it off in the Notification settings. I noticed they were not checked because I was looking for any reason why this wasn't working.

I turned it on by hand and that made it so it is successfully adding an image tracking beacon and links. I can confirm clicks and opens are being tracked in Cloudwatch:

<img width="1196" alt="Screen Shot 2020-10-16 at 6 09 54 AM" src="https://user-images.githubusercontent.com/6514609/96251559-45690480-0f76-11eb-8ef7-2edc3686e5ff.png">

Unfortunately, none of this is making it back to our dashboard:

<img width="517" alt="Screen Shot 2020-10-16 at 6 11 29 AM" src="https://user-images.githubusercontent.com/6514609/96251710-76493980-0f76-11eb-857f-3ae35369bf23.png">

Clearly SES knows about my SNS notification settings, as telling it to track clicks and opens in the AWS panel fixed that issue.
And we're getting emails.  But something is still misfiring here.

A second issue is we got one more webhook, and it's still not being processed, and I know for sure Horizon is running now.

**Comment by erikwestlund** [comment]
A couple more things:

I'm actually unsure whether or not I'm supposed to use "Track Clicks" and "Track Opens" in the SES settings:

![Screen Shot 2020-10-16 at 9 57 53 AM](https://user-images.githubusercontent.com/6514609/96274438-1a8ea880-0f96-11eb-9f81-65a20f2e1959.png)

As I noted above, I thought perhaps it was programmatically setting this, but as I think about this I do not believe there is a way for Mailcoach to do this with AWS access. **Should the docs reflect this needs to be set to on?**

For other drivers, it appears these are turned on.

When I have this on, as noted above, clicks get tracked by AWS.  

I've confirmed that we are getting SES feedback in the webhooks.

I see records in the DB (an example in above), and in NGINX I can see them come through:

[16/Oct/2020:15:05:14 +0000] 72.21.217.141 - - "POST /ses-feedback HTTP/1.1" 200 47 "-" "Amazon Simple Notification Service Agent" "5e32ad49c99d0cb1-EWR"

But even that webhook is not being processed.  They just sit.  I can confirm a job is getting run:

![Screen Shot 2020-10-16 at 9 56 58 AM](https://user-images.githubusercontent.com/6514609/96274889-a7d1fd00-0f96-11eb-9247-5a2d7716e9d8.png)

But nothing gets processed.  I'm guessing it doesn't process the Subscription Confirmed events?

Finally, I do not believe clicked links are sending any webhooks at all. I monitored nginx traffic after clicking a link and do not see anything come into the /ses-feedback endpoints. The opens/clicks are getting processed in Cloudfront, but it doesn't appear that clicks are being sent to the /ses-feedback endpoints. I have been hunting through SES/config but I believe everything is set up correctly.

**Comment by freekmurze** [comment]
You'll need to check `send` at Amazon as well. We'll update that screenshot soon. Thanks for bringing this to our attention.

> But even that webhook is not being processed. They just sit. I can confirm a job is getting run:
If a job just stays hanging in Horizon, that means that Horizon is not configured correctly. Make sure you configured workers that pick up jobs in that queue.

> I have been hunting through SES/config but I believe everything is set up correctly.
Our own integration is with SES is working, and we don't get reports from other clients that it is not working. Could you please go through the settings once again? If it still fails, contact AWS suppor, because you definitely should get the webhooks delivered to your own endpoint.

**Comment by erikwestlund** [comment]
Thanks.

Could someone on staff confirm with SES confirm that they’re receiving web hooks for clicks and opens and bounces? I’ve tried everything and checked settings and can still only see clicks and opens in Cloudwatch. I’m willing to troubleshoot with someone too.

**Comment by riasvdv** [comment]
So below is a screenshot of a working SES configuration set:

![image](https://user-images.githubusercontent.com/3626559/96442198-03a4bc00-120b-11eb-941d-9e5dd63853cd.png)

You don't need to check `send`, but you do need to check `open` & `click` if you want to track those. Make sure the configuration set name is configured correcly (as far as I know this is case sensitive, the best thing to do is keep it all lowercase).

Also make sure a topic for SNS is created.

Next, you need to make sure your SNS (Simple Notification Service) subscription is confirmed:

![image](https://user-images.githubusercontent.com/3626559/96442419-61d19f00-120b-11eb-962d-e28f44127872.png)

If it's not confirmed, try re-adding it again, AWS does not send webhooks to unconfirmed subscriptions.

**Comment by erikwestlund** [comment]
That’s the settings I have.

My sns topic is confirmed. But I’m not getting the web hooks from opens,
with the exception of the web hook saying my subscription is confirmed.


Would you mind sharing your redacted privacy policies on sns and ses?

**Comment by riasvdv** [comment]
What privacy policies? Everything else is default, we don't do anything extra that isn't described in the docs here https://mailcoach.app/docs/v3/package/handling-feedback/amazon-ses

**Comment by erikwestlund** [comment]
Here are screenshots of all relevant configurations:

SES configuration set:

<img width="772" alt="Screen Shot 2020-10-19 at 8 38 57 AM" src="https://user-images.githubusercontent.com/6514609/96458488-98e18980-11e6-11eb-8be5-b1556f4ded84.png">

SNS Topics:

<img width="1098" alt="Screen Shot 2020-10-19 at 8 42 28 AM" src="https://user-images.githubusercontent.com/6514609/96458919-1a391c00-11e7-11eb-8667-cdefa0e402d3.png">

SNS Subscription:

<img width="1086" alt="Screen Shot 2020-10-19 at 8 40 05 AM" src="https://user-images.githubusercontent.com/6514609/96458822-ff66a780-11e6-11eb-8f60-66682ec96681.png">


My sns topic is confirmed. But I’m not getting the web hooks from opens,
with the exception of the web hook saying my subscription is confirmed.

Topic access policy:

```
{
  "Version": "2008-10-17",
  "Id": "__default_policy_ID",
  "Statement": [
    {
      "Sid": "__default_statement_ID",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": [
        "SNS:GetTopicAttributes",
        "SNS:SetTopicAttributes",
        "SNS:AddPermission",
        "SNS:RemovePermission",
        "SNS:DeleteTopic",
        "SNS:Subscribe",
        "SNS:ListSubscriptionsByTopic",
        "SNS:Publish",
        "SNS:Receive"
      ],
      "Resource": "arn:aws:sns:us-east-1:XXXXX:mailcoach",
      "Condition": {
        "StringEquals": {
          "AWS:SourceOwner": "XXXXX"
        }
      }
    }
  ]
}
```


Again, here's the successful parts of my install:

* Emails send fine
* Emails arrive with open image beacon and click-tracking links substituted by SES
* Amazon catches opens and click-tracking through Cloudwatch
* Webhooks are being ingressed (I have records in my database of susbcription confirmations, as well as test messages I've sent to the topic in the AWS console.)

Here is what isn't working:
* When users open emails or click links no webhook is sent to Mailcoach through webhooks
* We thus have no campaign- or user-specific data on opens, clicks, or bounces, etc.

I have confirmed nothing is hitting the webhook input by tailing all logs in nginx, grepping the endpoint. The subscription confirmations and tests come in, but nothing else.

For reference, here's a dashboard I made to track opens/clicks in Cloudwatch, to show it is tracking them but not sending webhooks:

<img width="943" alt="Screen Shot 2020-10-19 at 8 46 59 AM" src="https://user-images.githubusercontent.com/6514609/96459831-2ec9e400-11e8-11eb-9ce3-739614f6b3ea.png">

And here's mailcoach page showing successful sends, but no data except unsubscribes:

<img width="1006" alt="Screen Shot 2020-10-19 at 8 50 21 AM" src="https://user-images.githubusercontent.com/6514609/96459851-34bfc500-11e8-11eb-8c5f-57c1c24faa97.png">

**Comment by erikwestlund** [comment]
Figured it out. @freekmurze, you all may want to put a warning for others:

Cloudflare has recently stepped-up some of its Web Access Firewall (WAF) settings, and requests with JSON in them are getting flagged at a high rate. This especially affects Livewire parts of my app, to the point I've had to disable security on certain endpoints.

**Cloudflare was blocking all webhooks to my feedback endpoint coming from AWS**

I used a Page Rule to disable security on that endpoint and opens and clicks are now tracking correctly.

Mad at myself for not making this connection earlier, when I had been debugging a similar issue in UIs just last week.
